### ISSUE TITLE
Federation hangs requesting time

### ISSUE THREAD

**Issue by wfvining-snl** [GMLC-TDC/HELICS#2574]
**Describe the bug**
Federation appears to hang while requesting next time. All of the federates have requested a next time, and the logs show the following message(s) continuously:

```
...
[2023-11-09 12:00:05.464] [console] [trace] S814 (131074)[1808]::processing command time_request:From (1) Time(1809, 1809, 1809) to (131074)
[2023-11-09 12:00:05.464] [console] [trace] S814 (131074)[1808]::processing command time_request:From (131075) Time(1809, 2708, 1809) to (131074)
[2023-11-09 12:00:05.464] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 1809, 1809) to (1) from 131074
[2023-11-09 12:00:05.464] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 1809, 1809) to (131074) from 131074
[2023-11-09 12:00:05.464] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 9223372036.854776, 9223372036.854776) to (131075) from 131074
[2023-11-09 12:00:05.464] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 1809, 1809) to (131075) from 131074
[2023-11-09 12:00:05.464] [console] [trace] S814 (131074)[1808]::processing command time_request:From (131074) Time(1809, 1809, 1809) to (131074)
[2023-11-09 12:00:05.465] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (1) Time(1809, 1809, 1809) to (131074) from 1
[2023-11-09 12:00:05.465] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131075) Time(1809, 2708, 1809) to (131074) from 131075
[2023-11-09 12:00:05.465] [console] [trace] S814 (131074)[1808]::processing command time_request:From (1) Time(1809, 1809, 1809) to (131074)
[2023-11-09 12:00:05.465] [console] [trace] S814 (131074)[1808]::processing command time_request:From (131075) Time(1809, 2708, 1809) to (131074)
[2023-11-09 12:00:05.465] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 1809, 1809) to (1) from 131074
[2023-11-09 12:00:05.465] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 1809, 1809) to (131075) from 131074
[2023-11-09 12:00:05.465] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 9223372036.854776, 9223372036.854776) to (131074) from 131074
[2023-11-09 12:00:05.465] [console] [trace] S814_core (1879048194)[t=1808]::|| cmd:time_request:From (131074) Time(1809, 1809, 1809) to (131075) from 131074
[2023-11-09 12:00:05.465] [console] [trace] S814 (131074)[1808]::processing command time_request:From (131074) Time(1809, 9223372036.854776, 9223372036.854776) to (131074)
...
```

Based on the logs it looks like the amount of time taken to process each time request increases with each time grant:
```
$ grep "\[t=1\]::|| cmd:time_request" grid.log | wc -l
51
$ grep "\[t=2\]::|| cmd:time_request" grid.log | wc -l
104
$ grep "\[t=3\]::|| cmd:time_request" grid.log | wc -l
72
$ grep "\[t=4\]::|| cmd:time_request" grid.log | wc -l
72753
$ grep "\[t=904\]::|| cmd:time_request" grid.log | wc -l
23102
$ grep "\[t=905\]::|| cmd:time_request" grid.log | wc -l
41955
$ grep "\[t=906\]::|| cmd:time_request" grid.log | wc -l
98845
$ grep "\[t=907\]::|| cmd:time_request" grid.log | wc -l
141920
$ grep "\[t=908\]::|| cmd:time_request" grid.log | wc -l
1054678
```

**What is the expected behavior?**
A time should be granted to the federates. The federation works (doesn't hang) with an older version of HELICS (3.1.2.post8).

**To Reproduce**
The behavior occurs in a federation with 2 federates configured as follows:

```json
{
  "name": "grid",
  "core_name": "grid_core",
  "log_level": "trace",
  "core_type": "zmq",
  "period": 1.0,
  "subscriptions": [],
  "publications": [
    {
      "key": "total_power",
      "type":  "complex",
      "units": "kW"
    }
  ],
  "endpoints": [
    {
      "name": "reliability"
    },
    {
      "name": "metrics"
    }
  ]
}
```
The grid federate also configures the following subscriptions and publications for each storage device:
```python
        self._power_sub = federate.register_subscription(
            f"{device.name}/power", units="kW"
        )
        self._power_sub.set_default(complex(0.0, 0.0))
        self._voltage_pub = federate.register_publication(
            f"storage.{device.name}.voltage",
            kind=HelicsDataType.DOUBLE,
            units="pu"
        )
        self._soc_pub = federate.register_publication(
            f"storage.{device.name}.soc",
            kind=HelicsDataType.DOUBLE,
            units=""
        )
        self._power_pub = federate.register_publication(
            f"storage.{device.name}.power",
            kind=HelicsDataType.COMPLEX,
            units="kW"
        )
```

storage federate
```json
{
  "log_level": "trace",
  "core_type": "zmq",
  "subscriptions": [],
  "publications": [
    {
      "key": "power",
      "unit": "kW",
      "type": "complex"
    }
  ]
}
```

**Environment (please complete the following information):**

- Operating System: Windows
- Language Extension: Python
- what compiler or setup process did you use: `pip install helics`
- HELICS version: 

```bash
$ helics --version
helics, version v3.4.0

Python HELICS version v3.4.0

HELICS Library version 3.4.0 (2023-01-19)
```

**Additional context and information**
The increasing number of log statements for each time request makes me think there is some kind of resource leak and that _eventually_ a new time would be granted.

**Comment by trevorhardy** [comment]
Is there a place you can post the code so that we could try to correlate the error on our side?

The increasing time for each time grant is most likely something outside the HELICS API (at least it has been every time I've seen that in my own federations); are there any likely places to look?

We've had problems with federations hanging in the past and they've usually had to do with making a time request for helicsTimeMaxtime and relying on incoming publications or messages to wake up the federate. Are you doing anything like that? (It looks like you are based on your log files but I want to make sure.)

**Comment by wfvining-snl** [comment]
Yes, the storage federate does make requests for `helics_time_maxtime` and rely on input from the other federate to wake it up. But the "grid" federate makes requests for times less than the maxtime.

I'm not sure where I would look for something outside the HELICS API. I've added logging on either side of the call to `Federate.request_time()` and it simply never returns. What should I be looking for that might cause this?

Unfortunately I can't post the code right now. I am trying to make a smaller reproducer that I can share.

**Comment by trevorhardy** [comment]
Regarding the federation hanging, which federate is making the request to `helics_time_maxtime`?

**Comment by wfvining-snl** [comment]
The one I called "storage federate" above.

**Comment by phlptp** [comment]
Is this always at a particular time step or is it variable which time step it hangs at?

**Comment by wfvining-snl** [comment]
It is variable

**Comment by trevorhardy** [comment]
is it possible to remove the endpoint from the grid federate without breaking the federation? If possible, try running without it and see if it hangs. My experience has been endpoints are more trouble-some when using `helics_time_maxtime`.

**Comment by phlptp** [comment]
one possibility is to use the aysnc call to request time,  then checking if is has completed in a loop, and if it takes too long,  make a query to "root" for "global_time_debugging".  The result sometime gives some additional information that can be used for diagnostics.

**Comment by phlptp** [comment]
Also like Trevor said there are few possible corner cases with endpoints in the timing with max time.   Using a targeted endpoint might resolve them as well, if your communication is always going on the same path.

**Comment by wfvining-snl** [comment]
Okay, I will try to eliminate the endpoints and see if that fixes it. When you say targeted you mean set the "destination" for the endpoint?

**Comment by phlptp** [comment]
Yeah that might work.  Depends if you are defining them through the API or a config file

**Comment by wfvining-snl** [comment]
I missed that there is actually an endpoint in the "storage federate" (configured via the API) that is currently unused (I write messages to it, but nothing is listening for them). Eliminating it completely resolved the issue.

**Comment by trevorhardy** [comment]
If you're willing, it would still be helpful to have a stripped-down version of the code to help us with finding the bug in HELICS. HELICS shouldn't hang a federation under the conditions you've described so this is still a bug we need to fix.

**Comment by wfvining-snl** [comment]
Yeah, I'm going to keep working on my small example. At some point I am going to need/use that endpoint. If I get it to reproduce the problem I will update this issue (unless you'd rather close it and I can make a new one later).

**Comment by trevorhardy** [comment]
Let's keep it all here; thanks!

**Comment by phlptp** [comment]
We have had a few reports of things like this before,  they have all been resolvable with some minor tweaks or traceable to a problem with the federate itself.  But there is likely some missed edge condition in the timing system that we are not handing correctly in HELICS, but reproducing the right conditions to be able to debug them in a compact repeatable way has thus far been very elusive.

**Comment by trevorhardy** [comment]
Any further update on this?

**Comment by wfvining-snl** [comment]
@trevorhardy after your comments above regarding issues with requesting `helics_time_maxtime` I have taken to just avoiding it by passing in the maximum time I intend to run the simulation to each federate and requesting that time instead of `helics_time_maxtime`. I've had this problem come up in at least two other co-simulations (one with just HELICS Inputs, no Endpoints) and that fix has seemed to work every time.

I think there is still something weird happening inside HELICS. Unfortunately I don't have a nice simple reproducer still. With this work around, and given that a simple reproducer is hard to make, I'm fine if you want to close this issue.

**Comment by trevorhardy** [comment]
@phlptp, I'm seeing this today on the [Fundamental Integration User Guide Example](https://github.com/GMLC-TDC/HELICS-Examples/tree/main/user_guide_examples/fundamental/fundamental_integration). When I'm destroying the federate, I ask for MAXTIME to make sure I clear out any pending messages and the federates hang. When I request MAXTIME-1, it doesn't.
